posts/2020-05-15-five-years-of-rust.md

```diff
@@ -0,0 +1,497 @@
+expression?.method()?;           // New
+```
+
+**[1.14] — Rustup 1.0** Rustup is Rust's Toolchain manager, it allows you to
+seamlessly use any version of Rust or any of its tooling. What started as a
+humble shell script has become what the maintainers affectionately call a
+*"chimera"*. Being able to provide first class compiler version management across
+Linux, macOS, Windows, and the dozens of target platforms would have been a
+myth just five years ago.
+
+#### 2017
+
+**[1.15] — Derive Procedural Macros** Derive Macros allow you to create powerful
+and extensive strongly typed APIs without all the boilerplate. This was the
+first version of Rust you could use libraries like `serde` or `diesel`'s
+derive macros on stable.
+
+**[1.17] — Rustbuild** One of the biggest improvements for our contributors to
+the language was moving our build system from the initial `make` based system
+to using cargo. This has opened up `rust-lang/rust` to being a lot easier for
+members and newcomers alike to build and contribute to the project.
+
+**[1.20] — Associated Constants** Previously constants could only be associated
+with a module. In 1.20 we stabilised associating constants on struct, enums,
+and importantly traits. Making it easier to add rich sets of preset values for
+types in your API, such as common IP addresses or interesting numbers.
+
+#### 2018
+
+**[1.24] — Incremental Compilation** Before 1.24 when you made a change in your
+library rustc would have to re-compile all of the code. Now rustc is a lot
+smarter about caching as much as possible and only needing to re-generate
+what's needed.
+
+**[1.26] — Existential Types** The addition of `impl Trait` gives you expressive
+dynamic APIs with the benefits and performance of static dispatch.
+
+**[1.28] — Global Allocators** Previously you were restricted to using the
+allocator that rust provided. With the global allocator API you can now
+customise your allocator to one that suits your needs. This was an important
+step in enabling the creation of the `alloc` library, another subset of the
+standard library containing only the parts of std that need an allocator like
+`Vec` or `String`. Now it's easier than ever to use even more parts of the
+standard library on a variety of systems.
+
+**[1.31] — 2018 edition** The release of the 2018 edition was easily our biggest
+release since 1.0, adding a collection of syntax changes and improvements to
+writing Rust written in a completely backwards compatible fashion, allowing
+libraries built with different editions to seamlessly work together.
+
+- **Non-Lexical Lifetimes** A huge improvement to Rust's borrow checker,
+  allowing it to accept more verifiable safe code.
+- **Module System Improvements** Large UX improvements to how we define and
+  use modules.
+- **Const Functions** Const functions allow you to run and evaluate Rust code
+  at compile time.
+- **Rustfmt** A new code formatting tool built specifically for Rust.
```

Comment:
Can we call these rustfmt and clippy 1.0? The tools themselves had been around for a while before.